### PR TITLE
ExecuteDelete: Always generate Any form

### LINKED ISSUE #28524: Exception when translating GroupBy-FirstOrDefault-Contains on entity
```C#
    [ConditionalTheory]
    [MemberData(nameof(IsAsyncData))]
    public virtual Task Where_delete_predicate_with_group_by_aggregate_2(bool async)
        => AssertDelete(
            async,
            ss => ss.Set<OrderDetail>()
                .Where(e => ss.Set<Order>()
                            .GroupBy(o => o.CustomerID)
                            .Where(g => g.Count() > 9)
                            .Select(g => g.First()).Contains(e.Order)),
            rowsAffectedCount: 40);
```

Not a bulk delete issue but the where predicate fails to translate since Contains is converted to Any and we don't process `ShapedQueryExpression` properly.

### LINKED ISSUE #28745: ExecuteDelete with optional reference navigation in filter fails with SQLite
Same code passes with SQL Server.

Query:

```C#
context.Posts.Where(p => p.Blog!.Title.StartsWith("Arthur")).ExecuteDelete();
```

Full repro:

```C#
public class Blog
{
    public int Id { get; set; }
    public string? Title { get; set; }

    public virtual ICollection<Post> Posts { get; } = new List<Post>();
}

public class Post
{
    public int Id { get; set; }
    public virtual Blog? Blog { get; set; }
}

public class SomeDbContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlite("Data Source=test.db")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();

    public DbSet<Blog> Blogs => Set<Blog>();
    public DbSet<Post> Posts => Set<Post>();
}

public class Program
{
    public static void Main()
    {
        using (var context = new SomeDbContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();
        
            context.Posts.Where(p => p.Blog!.Title.StartsWith("Arthur")).ExecuteDelete();
        }
    }
}
```

Exception:

```
Unhandled exception. System.InvalidOperationException: The LINQ expression 'ShapedQueryExpression:
    QueryExpression:
        Projection Mapping:
            EmptyProjectionMember -> EntityProjectionExpression: Post
        SELECT 1
        FROM Posts AS p
        LEFT JOIN Blogs AS b ON p.BlogId == b.Id
        WHERE (b.Title != NULL) && (('Arthur' != NULL) && (b.Title LIKE 'Arthur%'))
    ShaperExpression: EntityShaperExpression:
            Post
            ValueBufferExpression:
                ProjectionBindingExpression: EmptyProjectionMember
            IsNullable: False

    .Contains(EntityShaperExpression:
        Post
        ValueBufferExpression:
            ProjectionBindingExpression: EmptyProjectionMember
        IsNullable: False
    )' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.<VisitMethodCall>g__CheckTranslated|15_0(ShapedQueryExpression translated, <>c__DisplayClass15_0& )
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.TranslateSubquery(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.TranslateInternal(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.Translate(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateExpression(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateLambdaExpression(ShapedQueryExpression shapedQueryExpression, LambdaExpression lambdaExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateWhere(ShapedQueryExpression source, LambdaExpression predicate)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateExecuteDelete(ShapedQueryExpression source)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.RelationalQueryableExtensions.ExecuteDelete[TSource](IQueryable`1 source)
   at Program.Main() in C:\local\code\AllTogetherNow\Daily\Daily.cs:line 66
```

### LINKED ISSUE #28752: Bind member on result of single entity result subquery
```C#
    [ConditionalTheory]
    [MemberData(nameof(IsAsyncData))]
    public virtual Task Update_Where_Join_set_property_from_joined_table(bool async)
    => AssertUpdate(
        async,
        ss => from c in ss.Set<Customer>().Where(c => c.CustomerID.StartsWith("F"))
              select new { c, LastOrder = c.Orders.OrderByDescending(o => o.OrderDate).FirstOrDefault() },
        e => e.c,
        s => s.SetProperty(c => c.c.City, c => c.LastOrder.OrderDate.ToString()),
        rowsAffectedCount: 35,
        (b, a) => Assert.All(a, c => Assert.NotNull(c.City)));
```
We fail to translate `c.LastOrder.OrderDate` because we don't correctly process the `ShapedQueryExpression` from `c.LastOrder` to recognize it as entity reference in SqlTranslator. 

Could affect #28524

### PR DESCRIPTION
**Pull request by smitpatel**
Visit retrived ProjectionBinding from select expression as projection binding can bind to non-SqlExpression too.

Resolves #28524
Resolves #28745
Resolves #28752

### CODE REVIEW

File: test/EFCore.Relational.Specification.Tests/BulkUpdates/FiltersInheritanceBulkUpdatesTestBase.cs [hunk 1]
```diff
@@ -53,6 +53,35 @@ public virtual Task Delete_where_using_hierarchy_derived(bool async)
...
+                    .Select(g => g.First()),
+            rowsAffectedCount: 1);
+
+    [ConditionalTheory(Skip = "Issue#26753")]
+    [MemberData(nameof(IsAsyncData))]
+    public virtual Task Delete_GroupBy_Where_Select_First_2(bool async)
+        => AssertDelete(
+            async,
+            ss => ss.Set<Animal>().Where(e => e == ss.Set<Animal>().GroupBy(e => e.CountryId)
+                                                .Where(g => g.Count() < 3).Select(g => g.First()).FirstOrDefault()),
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/BulkUpdates/FiltersInheritanceBulkUpdatesTestBase.cs:73]: Indentation (also below) and in the other files

**Review by roji — APPROVED** [review]
Understanding a bit less but that is not preventing me from pushing approve!

:sheep: :it:

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -739,8 +812,8 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
             var item = Visit(arguments[1]);
 
             if (TryRewriteContainsEntity(
-                    enumerable,
-                    item == QueryCompilationContext.NotTranslatedExpression ? arguments[1] : item, out var result))
+                    enumerable == QueryCompilationContext.NotTranslatedExpression ? arguments[0] : enumerable,
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs:815]: revert this.

**Comment by smitpatel** [comment]
Majorly there are 2 changes here
- We moved processing of ShapedQueryExpression from VisitMethodCall to VisitExtension. This has moved back & forth few times. The bottom line is, VisitMethodCall only works if we send expression from outside which contains ShapedQueryExpression which is composed upon using queryable method call which would be translated as subquery. In any other case it doesn't work but that is not required limitation. Second point makes it even more obvious.
- Now we visit the projection we get from SelectExpression. A mapped projection can also be subquery (which can be lifted), (potentially JSON) so we need to visit it.

File: test/EFCore.Relational.Specification.Tests/BulkUpdates/FiltersInheritanceBulkUpdatesTestBase.cs (hunk 1 above)
Comment (smitpatel) [reply test/EFCore.Relational.Specification.Tests/BulkUpdates/FiltersInheritanceBulkUpdatesTestBase.cs:73]: Test, aligned with the inner queryable to make it clearer.

**Comment by roji** [comment]
Thanks for the explanation @smitpatel

**Pull request merged**